src/liballoc/string.rs

```diff
@@ -647,6 +648,30 @@ impl String {
         decode_utf16(v.iter().cloned()).map(|r| r.unwrap_or(REPLACEMENT_CHARACTER)).collect()
     }
 
+    /// Decomposes a `String` into its raw components.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(vec_into_raw_parts)]
+    /// let s = String::from("hello");
+    ///
+    /// let (ptr, len, cap) = s.into_raw_parts();
+    ///
+    /// let rebuilt = unsafe { String::from_raw_parts(ptr, len, cap) };
+    /// assert_eq!(rebuilt, "hello");
+    /// ```
+    #[unstable(feature = "vec_into_raw_parts", reason = "new API", issue="0")]
+    pub fn into_raw_parts(self) -> (*mut u8, usize, usize) {
+        let ptr = self.as_ptr() as *mut u8;
+        let len = self.len();
+        let cap = self.capacity();
+
+        mem::forget(self);
```

Comment:
The usually better style is to wrap the thing-to-forget in `ManuallyDrop` and then get its interior pointers. With these types and for current Stacked Borrows that makes no difference, but you never know.
```rust
        let s = ManuallyDrop::new(s);
        let ptr = s.as_ptr() as *mut u8;
        let len = s.len();
        let cap = s.capacity();
```

Reply:
Hmm. Doesn't such a style make it easier to accidentally use an aliased `s`? `mem::forget` removes the ability to use `s` after that point.

Reply:
The reason for this style is that the aliasing model *might* make it a meaningful operation to just pass a reference to another function -- that asserts the reference is live and all its uniqueness properties hold. `mem::forget` is a function like any other.

So, this might be UB (it is under Stacked Borrows):
```rust
let b = Box::new(0);
let raw = &mut *b;
mem::forget(b); // asserts uniqueness of the box!
let _val = *raw; // UB
```

> Doesn't such a style make it easier to accidentally use an aliased s? 

If you are worried about later re-use of a variable you can always shadow it:
```rust
        let s = ManuallyDrop::new(s);
        let ptr = s.as_ptr() as *mut u8;
        let len = s.len();
        let cap = s.capacity();
        let s; // shadow s
```
For short functions like the one here that seems like overdoing it, though.

Reply:
> you can always shadow it:

Not quite:

```
error[E0282]: type annotations needed
 --> src/main.rs:6:5
  |
6 | let s; // shadow s
  |     ^
  |     |
  |     cannot infer type
  |     consider giving `s` a type
```

I'd probably just... `drop` the `ManuallyDrop` in those cases:

```rust
let s = String::new();
let s = ManuallyDrop::new(s);
drop(s);
```

Reply:
Should I use `ManuallyDrop` in the documentation for the existing methods as well?

Reply:
> I'd probably just... drop the ManuallyDrop in those cases:

That passes it to a function again though.^^ Maybe wrapping it in a `ManuallyDrop` makes a difference, but until we know I'd advise against touching it.

You can shadow via `let s = ();` instead.

> Should I use ManuallyDrop in the documentation for the existing methods as well?

I think that would be better, yes. If you would have time to edit them that'd be awesome :)